File: 2-String/645_reverseVowels.py
import re
class Solution:
#     def reverseVowels1(self, s):
#         """
#         :type s: str
#         :rtype: str
#         """
#         temp = 'aeiouAEIOU'
#         l, r = 0, len(s) - 1
#         while l < r :
#             while l<r and s[l] not in temp:
#                 l += 1
#             while l<r and s[r] not in temp:
#                 r -= 1
#             if l < r:
#                 s = s[0:l] + s[r] + s[l+1:r] + s[l] + s[r+1:]
#             l += 1; r -= 1
#         return s
# #'str' object doesn't support item assignment  
    
    def reverseVowels(self, s):
        vowels = re.findall('(?i)[aeiou]', s)
        return re.sub('(?i)[aeiou]', lambda m: vowels.pop(), s)

sol = Solution()
s = 'aA'
ans = sol.reverseVowels(s)
print(ans)
#print : holle

File: 2-String/645_reverseVowels.java
public class Solution{
    public String reverseVowels(String s){
        if(s == null || s.length() == 0) reutrn s;
        String vowels = 'aeiouAEIOU';
        char[] chars = s.toCharArray();
        int start = 0;
        int end = s.length() - 1;
        while(start < end){
            while(start<end && !vowels.contains(chars[start]+"")){
                start ++;
            }
            while(start<end && !vowels.contains(chars[end]+"")){
                end --;
            }
            char temp = chars[start];
            chars[start] = chars[end];
            chars[end] = temp;

            start ++;
            end --;
        }
        return new String(chars);
    }

    public String reverseVowels1(String s){
        if(s == null || s.length() == 0）
            return s;
        int i = 0, j = s.length() - 1;
        char[] str = s.toCharArray();
        while(i<j){
            while(i<j && !isVowel(str[i]))
                i++;
            while(i<j && !isVowel(str[j]))
                j--;
            
            char temp = str[i];
            str[i] = str[j];
            str[j] = temp;
            i++;
            j--;
        }
        return new String(str);
    }
    private boolean isVowel(char ch){
        ch = Character.toLowerCase(ch);
        return ch == 'a' || ch == 'e' || ch == 'i' || ch == 'o' || ch == 'u';
    }
}